sklearn/linear_model/sag_fast.pyx

```diff
@@ -0,0 +1,341 @@
+                    cumulative_sums[itr] = (cumulative_sums[itr - 1]
+                                            + step_size / (wscale * num_seen))
+
+                # if wscale gets too small, we need to reset the scale
+                if wscale < 1e-9:
+                    if verbose:
+                        with gil:
+                            print("rescaling...")
+                    scale_weights(weights, wscale, n_features, n_samples,
+                                  itr, cumulative_sums, feature_hist,
+                                  sum_gradient)
+                    wscale = 1.0
+
+            if infinity:
+                break
+
+            # check if the stopping criteria is reached
+            scale_weights(weights, wscale, n_features, n_samples, n_samples - 1,
+                          cumulative_sums, feature_hist, sum_gradient)
+            wscale = 1.0
+            
+            max_change = 0.0
+            max_weight = 0.0
+            for j in range(n_features):
+                max_weight = fmax(max_weight, fabs(weights[j]))
+            for j in range(n_features):
+                max_change = fmax(max_change,
+                                  fabs(weights[j] - previous_weights[j]))
+                previous_weights[j] = weights[j]
+
+            n_iter += 1
+            if max_change / max_weight <= tol:
+                if verbose:
+                    end_time = time(NULL)
+                    with gil:
+                        print("convergence after %d epochs took %d seconds" %
+                              (n_iter, end_time - start_time))
+                break
+
+            if n_iter >= max_iter:
+                if verbose:
+                    end_time = time(NULL)
+                    with gil:
+                        print(("max_iter reached after %d seconds") %
+                              (end_time - start_time))
+                break
+
+    if infinity:
+        raise ValueError(("Floating-point under-/overflow occurred at epoch"
+                          " #%d. Lowering the step_size or scaling the input"
+                          " data with StandardScaler or"
+                          " MinMaxScaler might help.") % (n_iter + 1))
+
+
+    return intercept, num_seen, n_iter, intercept_sum_gradient
+
+
+cdef void scale_weights(double* weights, double wscale, int n_features,
+                        int n_samples, int itr, double* cumulative_sums,
+                        int* feature_hist, double* sum_gradient) nogil:
+    cdef int j
+    cdef double cum_sum
+    for j in range(n_features):
+        cum_sum = cumulative_sums[itr]
+        if feature_hist[j] != 0:
+            cum_sum -= cumulative_sums[feature_hist[j] - 1]
+
+        weights[j] -= cum_sum * sum_gradient[j]
+
+        weights[j] *= wscale
+        feature_hist[j] = (itr + 1) % n_samples
+
+    cumulative_sums[itr % n_samples] = 0.0
+
+
+cdef double dot(double* x_data_ptr, int* x_ind_ptr, double* w_data_ptr,
```

Comment:
should use the BLAS ddot function instead of defining another dot product function. Look inside linear_model/cd_fast.pyx or cluster/_k_means.pyx for example usage.